rasa/nlu/extractors/entity_synonyms.py

```diff
@@ -1,41 +1,65 @@
 import os
-from typing import Any, Dict, List, Optional, Text, Type
+from typing import Any, Dict, List, Optional, Text
 
-from rasa.nlu.components import Component
+from rasa.engine.graph import GraphComponent, ExecutionContext
 from rasa.shared.constants import DOCS_URL_TRAINING_DATA
 from rasa.shared.nlu.constants import ENTITIES, TEXT
-from rasa.nlu.config import RasaNLUModelConfig
-from rasa.nlu.extractors.extractor import EntityExtractor
-from rasa.nlu.model import Metadata
 from rasa.shared.nlu.training_data.training_data import TrainingData
 from rasa.shared.nlu.training_data.message import Message
 from rasa.nlu.utils import write_json_to_file
+from rasa.nlu.extractors.extractor import EntityExtractorMixin
 import rasa.utils.io
 import rasa.shared.utils.io
+from rasa.nlu.extractors._entity_synonyms import EntitySynonymMapper
+from rasa.engine.storage.resource import Resource
+from rasa.engine.storage.storage import ModelStorage
 
+# This is a workaround around until we have all components migrated to `GraphComponent`.
+EntitySynonymMapper = EntitySynonymMapper
 
-class EntitySynonymMapper(EntityExtractor):
-    @classmethod
-    def required_components(cls) -> List[Type[Component]]:
-        return [EntityExtractor]
+
+class EntitySynonymMapperComponent(GraphComponent, EntityExtractorMixin):
+    """Maps entities to their synonyms if they appear in the training data."""
+
+    SYNONYM_FILENAME = "synonyms.json"
 
     def __init__(
         self,
-        component_config: Optional[Dict[Text, Any]] = None,
+        config: Optional[Dict[Text, Any]] = None,
+        model_storage: ModelStorage = None,
+        resource: Resource = None,
         synonyms: Optional[Dict[Text, Any]] = None,
     ) -> None:
+        """Creates the mapper.
+
+        Args:
+            config: The mapper's config.
+            model_storage: Storage which the component can use to persist and load
+                itself.
+            resource: Resource locator for this component which can be used to persist
+                and load itself from the `model_storage`.
+            synonyms: A dictionary of previously known synonyms.
+        """
+        self._config = config
+        self._model_storage = model_storage
+        self._resource = resource
+
+        self._synonyms = synonyms if synonyms else {}
 
-        super().__init__(component_config)
-
-        self.synonyms = synonyms if synonyms else {}
-
-    def train(
-        self,
-        training_data: TrainingData,
-        config: Optional[RasaNLUModelConfig] = None,
-        **kwargs: Any,
-    ) -> None:
+    @classmethod
+    def create(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+        synonyms: Optional[Dict[Text, Any]] = None,
+    ) -> EntitySynonymMapperComponent:
```

Comment:
I seem to get a lot of errors about this not being defined yet locally - seems to really want me to use forward references, but not sure what I can do to avoid them like the other components?

Reply:
you need a `from __future__ import annotations` at the top of your file